### DISCUSSION TITLE
How to use Vercel previews and functions as GraphQL Endpoint?

### DISCUSSION THREAD

**Discussion by karltaylor** [vercel/vercel#6506] [Help]
I am deploying an application using Vercel, I am using previews.

When I make a PR, the app is built. ApolloClient on the frontend is using the environment variable `VERCEL_URL`.

However, when the `preview` is being built,  when ApolloClient tries to call the server but it does not return the correct data because the endpoint is not built yet. 

How do you get around this?

Here you can see the error, and the Vercel HTML source code when trying to query the Graphql Endpoint.

![Screenshot 2021-07-22 at 15 43 41](https://user-images.githubusercontent.com/6763109/126658637-53354df6-4e96-43f2-a238-ef83b665797c.png)

Edit: Related - https://github.com/vercel/next.js/discussions/16068#discussioncomment-126804

**Comment by leerob** [comment]
Rather than calling an API route from `getStaticProps` or `getServerSideProps`, you can move that logic directly into those functions, since they both run on the server-side, preventing this issue. This is also faster, since there's one less network hop needed.

> Note: You should not use fetch() to call an API route in getStaticProps. Instead, directly import the logic used inside your API route. You may need to slightly refactor your code for this approach.
> 
> Fetching from an external API is fine!

https://nextjs.org/docs/basic-features/data-fetching

**Reply by karltaylor** [reply]
Thanks for this Leerob, it's putting me in the right direction.

I came across [apollo-link-schema](https://www.apollographql.com/docs/react/api/link/apollo-link-schema/) which I think is the correct approach, however, I am unsure how to get my reference my schemas, as they're async. More of a GraphQL problem but thanks for your input

My ApolloSchema is a collection of subschemas:

```ts
export const makeSchemaAsync = async () => {
  const sanitySchema = await makeSanitySchema()
  const shopifySchema = await makeShopifySchema()

  return stitchSchemas({
    subschemas: [
      algoliaSchema,
      sanitySchema,
      shopifySchema,
    ],
  })
}

const makeApolloServerAsync = async () => {
  const apolloSever = new ApolloServer({
    schema: await makeSchemaAsync(),
    dataSources,
    introspection: true,
    playground: true,
  })

  return apolloSever
}

export default makeApolloServerAsync
```
